src/Features/Core/Portable/ExtractInterface/AbstractExtractInterfaceService.cs

```diff
@@ -147,50 +164,129 @@ public ExtractInterfaceResult ExtractInterfaceFromAnalyzedType(ExtractInterfaceT
         }
 
-        private Dictionary<ISymbol, SyntaxAnnotation> CreateSymbolToDeclarationAnnotationMap(
-            IEnumerable<ISymbol> includedMembers,
-            ref Solution solution,
-            out List<DocumentId> documentIds,
-            SyntaxNode typeNode,
-            out SyntaxAnnotation typeNodeAnnotation,
-            CancellationToken cancellationToken)
+        private async Task<ExtractInterfaceResult> ExtractInterfaceToSameFileAsync(
+            Solution solution, ExtractInterfaceTypeAnalysisResult refactoringResult, INamedTypeSymbol extractedInterfaceSymbol,
+            ExtractInterfaceOptionsResult extractInterfaceOptions, CancellationToken cancellationToken)
+        {
+            // Track all of the symbols we need to modify, which includes the original type declaration being modified
+            var symbolMapping = await CreateSymbolToDeclarationAnnotationMapAsync(
+                extractInterfaceOptions.IncludedMembers,
+                solution,
+                refactoringResult.TypeNode,
+                cancellationToken).ConfigureAwait(false);
+
+            var document = symbolMapping.AnnotatedSolution.GetDocument(refactoringResult.DocumentToExtractFrom.Id);
+            var originalRoot = document.GetSyntaxRootSynchronously(cancellationToken);
+            var typeDeclaration = originalRoot.GetAnnotatedNodes(symbolMapping.TypeNodeAnnotation).Single();
+
+            var trackedDocument = document.WithSyntaxRoot(originalRoot.TrackNodes(typeDeclaration));
+
+            var currentRoot = trackedDocument.GetSyntaxRootSynchronously(cancellationToken);
+            var editor = new SyntaxEditor(currentRoot, symbolMapping.AnnotatedSolution.Workspace);
+
+            // Generate the interface syntax node, which will be inserted above the type it's extracted from
+            var codeGenService = trackedDocument.GetLanguageService<ICodeGenerationService>();
+            var interfaceNode = codeGenService.CreateNamedTypeDeclaration(extractedInterfaceSymbol).WithAdditionalAnnotations(SimplificationHelpers.SimplifyModuleNameAnnotation);
+
+            typeDeclaration = currentRoot.GetCurrentNode(typeDeclaration);
+            editor.InsertBefore(typeDeclaration, interfaceNode);
+
+            var unformattedSolution = document.WithSyntaxRoot(editor.GetChangedRoot()).Project.Solution;
+
+            // After the interface is inserted, update the original type to show it implements the new interface
+            var completedSolution = await GetSolutionWithOriginalTypeUpdatedAsync(
+                unformattedSolution, symbolMapping.DocumentIds,
+                refactoringResult.DocumentToExtractFrom.Id, symbolMapping.TypeNodeAnnotation,
+                refactoringResult.TypeToExtractFrom, extractedInterfaceSymbol,
+                extractInterfaceOptions.IncludedMembers, symbolMapping.SymbolToDeclarationAnnotationMap, cancellationToken).ConfigureAwait(false);
```

Comment:
pass along symbolMapping instance to this helper?

Reply:
I prefer to leave as is, with the implementations being more clear on what information is being passed. `SymbolMapping` is specifically existing as an identity to encapsulate the work from a method, which may not make sense in this context. I don't care to increase the scope and usefulness of `SymbolMapping`, and without that I think it's better to use it only as a private type.